server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.composite;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.LongArray;
+import org.elasticsearch.common.util.ObjectArray;
+import org.elasticsearch.core.CheckedFunction;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.StringFieldType;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+
+import java.io.IOException;
+import java.util.function.LongConsumer;
+
+import static org.apache.lucene.index.SortedSetDocValues.NO_MORE_ORDS;
+
+/**
+ * A {@link SingleDimensionValuesSource} for ordinals.
+ */
+class OrdinalValuesSource extends SingleDimensionValuesSource<BytesRef> {
+    private final LongConsumer breakerConsumer;
+    private final CheckedFunction<LeafReaderContext, SortedSetDocValues, IOException> docValuesFunc;
+
+    // ordinals, which are remapped whenever we visit a new segment.
+    // Entries might be Long.MIN_VALUE to represent the missing bucket,
+    // or negative when the corresponding term is not known to the current lookup
+    private LongArray valuesOrd;
+    // when term is not known to current lookup, then the term from a previous lookup is stored here, else null
+    private ObjectArray<BytesRef> values;
+    // number of slots in the above arrays that contain any actual values
+    private int numSlots = 0;
+
+
+    // is Long.MIN_VALUE when the value represents the missing bucket, or negative if the term is not known to current lookup
+    private Long currentValueOrd;
+    // when term is not known to current lookup, then the term from a previous lookup is stored here, else null
```

Comment:
I think these comments are a little confusing.  Maybe have one block comment to explain how we're using `Long.MIN_VALUE`, negatives, and nulls here and then annotate the three different pairs of ordinals and values with their purpose?